# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Test DNA sequences for equality.
#' 
#' \code{rcpp_testSeqEqual} checks if two DNA sequences are identical.
#'
#' @param    seq1    character string containing a DNA sequence.
#' @param    seq2    character string containing a DNA sequence.
#' @param    ignore  vector of characters to ignore when testing for equality.
#' 
#' @return   Returns \code{TRUE} if sequences are equal and \code{FALSE} if they are not.
#'           Sequences of unequal length will always return \code{FALSE} regardless of
#'           their character values.
#' 
#' @seealso  Used by \link{collapseDuplicates}.
#' 
#' @examples
#' # Ignore gaps
#' rcpp_testSeqEqual("ATG-C", "AT--C")
#' rcpp_testSeqEqual("ATGGC", "ATGGN")
#' rcpp_testSeqEqual("AT--T", "ATGGC")
#' 
#' # Ignore only Ns
#' rcpp_testSeqEqual("ATG-C", "AT--C", ignore="N")
#' rcpp_testSeqEqual("ATGGC", "ATGGN", ignore="N")
#' rcpp_testSeqEqual("AT--T", "ATGGC", ignore="N")
#' 
#' @export
rcpp_testSeqEqual <- function(seq1, seq2, ignore = character()) {
    .Call('alakazam_rcpp_testSeqEqual', PACKAGE = 'alakazam', seq1, seq2, ignore)
}

#' Calculate pairwise equivalence between sequences
#' 
#' \code{getDistanceMatrix} determined pairwise equivalence between a pairs in a 
#' set of sequences, excluding ambiguous positions (Ns and gaps).
#'
#' @param    seq  character vector containing a DNA sequences.
#'
#' @return   A logical matrix of equivalence between each entry in \code{seq}. 
#'           Values are \code{TRUE} when sequences are equivalent and \code{FALSE}
#'           when they are not.
#' 
#' @seealso  Uses \link{testSeqEqual} for testing equivalence between pairs.
#'           
#' @examples
#' # Gaps and Ns will match any character
#' seq <- c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C", E="NTGGG")
#' d <- getDistanceMatrix(seq)
#' rownames(d) <- colnames(d) <- seq
#' d
#' 
#' @export
getDistanceMatrix <- function(seq) {
    .Call('alakazam_getDistanceMatrix', PACKAGE = 'alakazam', seq)
}

validChars <- function(seq1, seq2) {
    .Call('alakazam_validChars', PACKAGE = 'alakazam', seq1, seq2)
}

#' Calculate distance between two sequences
#' 
#' \code{getSeqDistance} calculates the distance between two DNA sequences.
#'
#' @param    seq1      character string containing a DNA sequence.
#' @param    seq2      character string containing a DNA sequence.
#' @param    dist_mat  Character distance matrix. Defaults to a Hamming distance 
#'                     matrix returned by \link{getDNAMatrix}. If gap 
#'                     characters, \code{c("-", ".")}, are assigned a value of -1 
#'                     in \code{dist_mat} then contiguous gaps of any run length,
#'                     which are not present in both sequences, will be counted as a 
#'                     distance of 1. Meaning, indels of any length will increase
#'                     the sequence distance by 1. Gap values other than -1 will 
#'                     return a distance that does not consider indels as a special case.
#'
#' @return   Numerical distance between \code{seq1} and \code{seq2}.
#' 
#' @seealso  Nucleotide distance matrix may be built with 
#'           \link{getDNAMatrix}. Amino acid distance matrix may be built
#'           with \link{getAAMatrix}.
#'           
#' @examples
#' # Ungapped examples
#' getSeqDistance("ATGGC", "ATGGG")
#' getSeqDistance("ATGGC", "ATG??")
#' 
#' # Gaps will be treated as Ns with a gap=0 distance matrix
#' getSeqDistance("ATGGC", "AT--C", dist_mat=getDNAMatrix(gap=0))
#' 
#' # Gaps will be treated as universally non-matching characters with gap=1
#' getSeqDistance("ATGGC", "AT--C", dist_mat=getDNAMatrix(gap=1))
#' 
#' # Gaps of any length will be treated as single mismatches with a gap=-1 distance matrix
#' getSeqDistance("ATGGC", "AT--C", dist_mat=getDNAMatrix(gap=-1))
#' 
#' # Gaps of equivalent run lengths are not counted as gaps
#' getSeqDistance("ATG-C", "ATG-C", dist_mat=getDNAMatrix(gap=-1))
#'
#' # Overlapping runs of gap characters are counted as a single gap
#' getSeqDistance("ATG-C", "AT--C", dist_mat=getDNAMatrix(gap=-1))
#' getSeqDistance("A-GGC", "AT--C", dist_mat=getDNAMatrix(gap=-1))
#' getSeqDistance("AT--C", "AT--C", dist_mat=getDNAMatrix(gap=-1))
#' 
#' # Discontiguous runs of gap characters each count as separate gaps
#' getSeqDistance("-TGGC", "AT--C", dist_mat=getDNAMatrix(gap=-1))
#' 
#' @export
rcpp_getSeqDistance <- function(seq1, seq2, dist_mat) {
    .Call('alakazam_rcpp_getSeqDistance', PACKAGE = 'alakazam', seq1, seq2, dist_mat)
}

#' Calculate pairwise distances between sequences
#' 
#' \code{getSeqMatrix} calculates all pairwise distance between a set of sequences.
#'
#' @param    seq       character vector containing a DNA sequences.
#' @param    dist_mat  Character distance matrix. Defaults to a Hamming distance 
#'                     matrix returned by \link{getDNAMatrix}. If gap 
#'                     characters, \code{c("-", ".")}, are assigned a value of -1 
#'                     in \code{dist_mat} then contiguous gaps of any run length,
#'                     which are not present in both sequences, will be counted as a 
#'                     distance of 1. Meaning, indels of any length will increase
#'                     the sequence distance by 1. Gap values other than -1 will 
#'                     return a distance that does not consider indels as a special case.
#'
#' @return   A matrix of numerical distance between each entry in \code{seq}. 
#'           If \code{seq} is a named vector, row and columns names will be added 
#'           accordingly.
#' 
#' @seealso  Uses \link{getSeqDistance} for calculating distances between pairs.
#'           Nucleotide distance matrix may be built with \link{getDNAMatrix}. 
#'           Amino acid distance matrix may be built with \link{getAAMatrix}. 
#'           
#' @examples
#' # Gaps will be treated as Ns with a gap=0 distance matrix
#' getSeqMatrix(c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C"), 
#'              dist_mat=getDNAMatrix(gap=0))
#' 
#' # Gaps will be treated as universally non-matching characters with gap=1
#' getSeqMatrix(c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C"), 
#'              dist_mat=getDNAMatrix(gap=1))
#' 
#' # Gaps of any length will be treated as single mismatches with a gap=-1 distance matrix
#' getSeqMatrix(c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C"), 
#'              dist_mat=getDNAMatrix(gap=-1))
#' 
#' @export
rcpp_getSeqMatrix <- function(seq, dist_mat) {
    .Call('alakazam_rcpp_getSeqMatrix', PACKAGE = 'alakazam', seq, dist_mat)
}

